lib/posix-process/tests/test_syscall_shim.c

```diff
@@ -0,0 +1,79 @@
+#include "test_syscall_shim.h"
+
+char TLS_REGION[100];
+unsigned long return_addr;
+int is_binary;
+
+unsigned long ktls;
+unsigned long ultls;
+unsigned long saved_return_addr;
+/*
+ * This syscall is only for testing the User land TLS.
+ * Not a real fork !!!
+ */
+UK_SYSCALL_R_DEFINE(int, fork)
+{
+	if (is_binary) {
+		struct uk_thread *self = uk_thread_current();
+
+		ktls = ukplat_tlsp_get();
+		if (self->uktlsp != ktls)
+			return 0;
+
+		if (ktls == (unsigned long) TLS_REGION)
+			return 0;
+
+		ultls = uk_syscall_ultlsp();
+	}
+
+	saved_return_addr = uk_syscall_return_addr();
+
+	return 0;
+}
+
+UK_TESTCASE(posix_process_syscall_shim_testsuite, test_fork)
+{
+	/* Save the real tls. */
+	unsigned long orig = ukplat_tlsp_get();
+
+	/* <imics an userland tls. */
+	ukplat_tlsp_set((unsigned long) TLS_REGION);
+
+	UK_TEST_ASSERTF(((void *) ukplat_tlsp_get() == TLS_REGION),
+		"expected ULTLS 0x%lx to be set to %p", ukplat_tlsp_get(),
+		TLS_REGION);
+
+	is_binary = 1;
+	call_binary_fork();
+
+	UK_TEST_ASSERTF((uk_thread_current()->uktlsp == ktls),
```

Comment:
When using musl this will make the program crash, because we don't have a valid TLS. I think we still need to do the checking (lines 49-67) but move all the printing after line 69, when the real TLS is restored.